### PR TITLE
Import export csv import configurations

### PR DESCRIPTION
**Pull request by MarcoGuiducci**
## Changes in this pull request  
Possibility to Import and Export the configurations created for CSV Imports

## Additional info  
Related Issue:
#6091

### CODE REVIEW

**Comment by weisswurstkanone** [comment]
please revert it.json
en.json is the only file needed . The rest is done in POEditor.
see https://poeditor.com/join/project/VWmZyvFVMH

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php [hunk 1]
```diff
@@ -2488,4 +2488,168 @@ protected function processAvailableFieldDefinitions($fds, &$firstOne, &$commonFi
...
+            throw new \Exception("Please provide a valid import configuration Id");
+        }
+
+        try {
+            $importConfig = ImportConfig::getById($importConfigId);
+        
+            $configData = $importConfig->getConfig();
+            $configName = $importConfig->getName();
+            
+            $response = new Response();
```
Comment (weisswurstkanone) [inline bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: better use a JsonResponse. Example: 
        $jsonResponse = new JsonResponse($json, 200, [
            'Content-Disposition' => 'attachment; filename="export-data-object-' . $object->getId() . '.json"'
        ], true);

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php [hunk 2]
```diff
@@ -2488,4 +2488,168 @@ protected function processAvailableFieldDefinitions($fds, &$firstOne, &$commonFi
...
+        $selectedGridColumns = $configData['selectedGridColumns'];
+        $resolverSettings = $configData['resolverSettings'];
+        $shareSettings = $configData['shareSettings'];
+        $dialect = json_decode(json_encode($configData['csvSettings']), false);
+        
+        $success = true;
+        $supportedFieldTypes = ['checkbox', 'country', 'date', 'datetime', 'href', 'image', 'input', 'language', 'table', 'multiselect', 'numeric', 'password', 'select', 'slider', 'textarea', 'wysiwyg', 'objects', 'multihref', 'geopoint', 'geopolygon', 'geopolyline', 'geobounds', 'link', 'user', 'email', 'gender', 'firstname', 'lastname', 'newsletterActive', 'newsletterConfirmed', 'countrymultiselect', 'objectsMetadata'];
+
+        $classId = $request->get('classId');
+        $file = PIMCORE_SYSTEM_TEMP_DIRECTORY . '/import_' . $request->get('importId');
```
Comment (weisswurstkanone) [inline bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: why are we dealing with csv files suddenly? my understanding is that we want to import the exported JSON configuration file?

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php [hunk 3]
```diff
@@ -2488,4 +2488,168 @@ protected function processAvailableFieldDefinitions($fds, &$firstOne, &$commonFi
...
+
+        $configData = json_decode($json, true);
+
+        $selectedGridColumns = $configData['selectedGridColumns'];
+        $resolverSettings = $configData['resolverSettings'];
+        $shareSettings = $configData['shareSettings'];
+        $dialect = json_decode(json_encode($configData['csvSettings']), false);
+        
+        $success = true;
+        $supportedFieldTypes = ['checkbox', 'country', 'date', 'datetime', 'href', 'image', 'input', 'language', 'table', 'multiselect', 'numeric', 'password', 'select', 'slider', 'textarea', 'wysiwyg', 'objects', 'multihref', 'geopoint', 'geopolygon', 'geopolyline', 'geobounds', 'link', 'user', 'email', 'gender', 'firstname', 'lastname', 'newsletterActive', 'newsletterConfirmed', 'countrymultiselect', 'objectsMetadata'];
```
Comment (weisswurstkanone) [inline bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: see question below but this shouldn't be hardcoded anyway. hard to maintain and even impossible to add custom datatypes.

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php (hunk 1 above)
Comment (MarcoGuiducci) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: Done

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php (hunk 2 above)
Comment (MarcoGuiducci) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: My idea was to produce the same JSON response of the importGetFileInfoAction function.
The difference is that the json configuration is taken from the uploaded file instead of being loaded from database

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php (hunk 3 above)
Comment (MarcoGuiducci) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: As said in the previous comment, this is part of the code of the importGetFileInfoAction. How can I modify to avoid this code duplication?

Comment (weisswurstkanone) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: well, importGetFileAction is about importing CSV data. Here we are talking about importing JSON configurations. Maybe take a look at 
https://github.com/pimcore/pimcore/blob/663b0887cf4623b5499987e2b15b9522f3342e11/bundles/AdminBundle/Controller/Admin/DataObject/ClassController.php#L460
for another sample as well.

I guess you don't need the whole magic.

Comment (MarcoGuiducci) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: My idea was to use the imported json to refresh the CSV import interface the same way it's done when a saved configuration is loaded. I will follow your suggestion and try to make this more simple

Comment (weisswurstkanone) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: Thank you! 👍

Comment (MarcoGuiducci) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: Made some changes.
As i wrote in the code's comments, I have maintained the reload of data form original CSV to properly refresh the data preview on the import interface

File: bundles/CoreBundle/Resources/translations/en.json [hunk 4]
```diff
@@ -942,5 +942,7 @@
   "domain_label_width": "Domain label width",
   "redirect_performance_warning": "Using this feature has very bad effects on the performance of the entire application, do you really want to continue?",
   "Uncategorized": "Uncategorized",
-  "methods": "Methods"
+  "methods": "Methods",
+  "export_configuration": "Export Configuration",
```
Comment (weisswurstkanone) [inline bundles/CoreBundle/Resources/translations/en.json]: maybe mention JSON here

something "export as json" and "import from json" just to avoid confusion with the other import

File: bundles/CoreBundle/Resources/translations/it.json [hunk 5]
```diff
@@ -940,4 +940,4 @@
     "test_url": "Test URL",
     "create_redirects": "Create redirects to keep URLs working (incl. children)?",
     "auto_create": "Auto create"
-}
```
Comment (weisswurstkanone) [inline bundles/CoreBundle/Resources/translations/it.json]: please remove entirely if possible

**Comment by weisswurstkanone** [comment]
can you resolve the conflict please?

in addition, please mention this feature in the docs
https://pimcore.com/docs/6.x/User_Documentation/DataObjects/CSV_Import/index.html

File: bundles/CoreBundle/Resources/translations/en.json (hunk 4 above)
Comment (MarcoGuiducci) [reply bundles/CoreBundle/Resources/translations/en.json]: Done, thanks for the suggestion

File: bundles/CoreBundle/Resources/translations/it.json (hunk 5 above)
Comment (MarcoGuiducci) [reply bundles/CoreBundle/Resources/translations/it.json]: I'm trying to revert this, but the blank line still remain

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php [hunk 6]
```diff
@@ -2489,4 +2489,119 @@ protected function processAvailableFieldDefinitions($fds, &$firstOne, &$commonFi
...
+     *
+     * @param Request $request
+     * 
+     * @return Response
+     */
+    public function exportJsonConfigAction(Request $request)
+    {
+        $importConfigId = $request->get('importConfigId');
+        
+        if($importConfigId == null || empty($importConfigId)){
```
Comment (NiklasBr) [inline bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: `if (!isset($importConfigId))` should be enough here because `isset(null)` is `false`

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php [hunk 7]
```diff
@@ -2489,4 +2489,119 @@ protected function processAvailableFieldDefinitions($fds, &$firstOne, &$commonFi
...
+                foreach ($rowData as $key => $value) {
+                    $tmpData['field_' . $key] = $value;
+                }
+
+                $tmpData['rowId'] = $count + 1;
+                $data[] = $tmpData;
+
+                $count++;
+
+                if ($count > 18) {
```
Comment (NiklasBr) [inline bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: `18` is a magic number of some sort, would be nice if it was described.

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php (hunk 6 above)
Comment (MarcoGuiducci) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: Done. Thanks for the suggestion

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php (hunk 7 above)
Comment (MarcoGuiducci) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: I have added a comment, but this was taken from originale Pimcore's code

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php [hunk 8]
```diff
@@ -2489,4 +2489,122 @@ protected function processAvailableFieldDefinitions($fds, &$firstOne, &$commonFi
...
+        
+        $success = true;
+        $classId = $request->get('classId');
+        
+        $file = PIMCORE_SYSTEM_TEMP_DIRECTORY . '/import_' . $request->get('importId');
+
+        $originalFile = $file . '_original';
+        // determine type
+        if (empty($dialect)) {
+            $dialect = Tool\Admin::determineCsvDialect($file . '_original');
```
Comment (NiklasBr) [inline bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: You have `$originalFile` above, it can be re-used rather than concatenating again.

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php [hunk 9]
```diff
@@ -2500,4 +2500,121 @@ protected function processAvailableFieldDefinitions($fds, &$firstOne, &$commonFi
...
+                 * Reached the number or rows for the preview
+                 */
+                if ($count > 18) {
+                    break;
+                }
+            }
+            fclose($handle);
+        }
+        
+        $availableConfigs = $this->getImportConfigs($importService, $this->getAdminUser(), $classId);
```
Comment (weisswurstkanone) [inline bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: why do we use the admin user here ? shouldn't it be the user that triggers the import action ?

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php [hunk 10]
```diff
@@ -2500,4 +2500,121 @@ protected function processAvailableFieldDefinitions($fds, &$firstOne, &$commonFi
...
+
+        return $this->adminJson([
+            'success' => $success,
+            'config' => [
+                'importConfigId' => $importConfigId,
```
Comment (weisswurstkanone) [inline bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: where does this importConfigId come from? I guess when yu import it from JSON it shouldn't be set. right now, would it be possible to overwrite a config I don't even own?

File: bundles/AdminBundle/Resources/public/js/pimcore/object/helpers/import/configDialog.js [hunk 11]
```diff
@@ -125,6 +125,14 @@ pimcore.object.helpers.import.configDialog = Class.create({
 
         buttons.push(this.deleteButton);
 
+        this.exportConfigButton = new Ext.menu.Item({
```
Comment (weisswurstkanone) [inline bundles/AdminBundle/Resources/public/js/pimcore/object/helpers/import/configDialog.js]: From a user experience perspective I believe that the "Load" split button is not the right place for exporting something. Maybe rephrase the entire menu? Or even combine "Load" and "Save" split button ?

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php [hunk 12]
```diff
@@ -2500,4 +2500,121 @@ protected function processAvailableFieldDefinitions($fds, &$firstOne, &$commonFi
...
+    
+    /**
+     * @Route("/import-json-config", methods={"POST"})
+     *
+     * @param Request $request
+     * @param ImportService $importService
+     *
+     * @return JsonResponse
+     */
+    public function importJsonConfigAction(Request $request, ImportService $importService)
```
Comment (weisswurstkanone) [inline bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: maybe change this to
importCSVImportConfigFromJsonAction or something simpler (same for route).

Currently it is not clear what this is all about :-)

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php (hunk 10 above)
Comment (weisswurstkanone) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: we could also try to share identical code. see importGetFileInfoAction for example (same controller)

Comment (MarcoGuiducci) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: the importConfigId comes from the requests, it corresponds to the current loaded configuration, it's not overridden by the imported JSON

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php (hunk 9 above)
Comment (MarcoGuiducci) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: I'm trying to use the $this->getUser() function, but i got this exception.

Exception

The exception is thrown on the $importService->getSharedImportConfigs($user, $classId) function.

Which is the correct function to call to get the current user?

Comment (weisswurstkanone) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: \Pimcore\Tool\Admin::getCurrentUser() I believe

Comment (MarcoGuiducci) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: Thank you, fixed

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php (hunk 10 above)
Comment (MarcoGuiducci) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: > we could also try to share identical code. see importGetFileInfoAction for example (same controller)

Done sharing code

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php (hunk 12 above)
Comment (MarcoGuiducci) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: route and action renamed ;)

File: bundles/AdminBundle/Resources/public/js/pimcore/object/helpers/import/configDialog.js (hunk 11 above)
Comment (MarcoGuiducci) [reply bundles/AdminBundle/Resources/public/js/pimcore/object/helpers/import/configDialog.js]: Sincerely, i'm not very skilled in terms of UX :)
I think that the "Import" button could remain in the Load split button. I can just move the "Export" behind the Save split button.
What do you think about it?

Comment (weisswurstkanone) [reply bundles/AdminBundle/Resources/public/js/pimcore/object/helpers/import/configDialog.js]: better than now, because you will never search for an export button in a "Load" menu :-)

Comment (MarcoGuiducci) [reply bundles/AdminBundle/Resources/public/js/pimcore/object/helpers/import/configDialog.js]: Moved the Export functionality on the Save split button

**Comment by weisswurstkanone** [comment]
Resolved all conversations? Ready to be reviewed again ? 🤔

**Comment by MarcoGuiducci** [comment]
> Resolved all conversations? Ready to be reviewed again ? 🤔

Yes, I think that I've done all the changes you suggested :)

File: bundles/AdminBundle/Resources/public/js/pimcore/object/helpers/import/configDialog.js [hunk 13]
```diff
@@ -773,5 +795,17 @@ pimcore.object.helpers.import.configDialog = Class.create({
...
             }.bind(this)
         });
+    },
+    
+    getExportConfigUrl: function(){
+        this.commitEverything();
+        var config = this.prepareSaveData();
+        config = Ext.encode(config);
+        
+        return "/admin/object-helper/export-json-config?classId="+this.classId+"&config="+config;
```
Comment (weisswurstkanone) [inline bundles/AdminBundle/Resources/public/js/pimcore/object/helpers/import/configDialog.js:805]: same similar to import url

File: bundles/CoreBundle/Resources/translations/en.json [hunk 14]
```diff
@@ -944,6 +944,8 @@
   "redirect_performance_warning": "Using this feature has very bad effects on the performance of the entire application, do you really want to continue?",
   "Uncategorized": "Uncategorized",
   "methods": "Methods",
+  "export_configuration": "Export As JSON",
```
Comment (weisswurstkanone) [inline bundles/CoreBundle/Resources/translations/en.json:947]: better Export Configuration as JSON
same for import

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php [hunk 15]
```diff
@@ -2500,4 +2511,98 @@ protected function processAvailableFieldDefinitions($fds, &$firstOne, &$commonFi
             }
         }
     }
+    
+    /**
+     * @Route("/export-json-config", methods={"GET"})
```
Comment (weisswurstkanone) [inline bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php:2516]: same as for import, maybe 
export-csv-import-config-from-json

File: bundles/AdminBundle/Resources/public/js/pimcore/object/helpers/import/configDialog.js (hunk 13 above)
Comment (MarcoGuiducci) [reply bundles/AdminBundle/Resources/public/js/pimcore/object/helpers/import/configDialog.js]: Done, renamed as "export-csv-import-config-as-json"

File: bundles/CoreBundle/Resources/translations/en.json (hunk 14 above)
Comment (MarcoGuiducci) [reply bundles/CoreBundle/Resources/translations/en.json]: Done

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php (hunk 15 above)
Comment (MarcoGuiducci) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: Done, renamed as "export-csv-import-config-as-json"

File: bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php (hunk 7 above)
Comment (weisswurstkanone) [reply bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php]: this is just a random number, simply for getting the first x number of lines as preview

**Comment by weisswurstkanone** [comment]
@MarcoGuiducci looks good!

One last wish: Could you please check the user documentation (https://pimcore.com/docs/6.x/User_Documentation/DataObjects/CSV_Import/index.html)
and modify/extend it if necessary?

**Comment by MarcoGuiducci** [comment]
> @MarcoGuiducci looks good!
> 
> One last wish: Could you please check the user documentation (https://pimcore.com/docs/6.x/User_Documentation/DataObjects/CSV_Import/index.html)
> and modify/extend it if necessary?

I've done it, but accidentally another PR was opened for that

**Comment by weisswurstkanone** [comment]
Thanks a lot, definitely a useful improvement!

**Comment by MarcoGuiducci** [comment]
> Thanks a lot, definitely a useful improvement!

It was my pleasure!!! Thanks for your suggestions during these days
